### ISSUE TITLE
Support image1 * image2 for different pixel types

### ISSUE THREAD

**Issue by dyollb** [SimpleITK/SimpleITK#1723]
Often it is convenient to multiply one image by another, e.g. to mask a segmentation or MRI.

Currently, SimpleITK rasies an exception if the pixel types are different. It would be nice if SimpleITK would automatically cast following similar rules as [standard conversions](https://learn.microsoft.com/en-us/cpp/cpp/standard-conversions?view=msvc-170) in C++, e.g. 
- `float * double ==> double`
- `float * int16 ==> float`
- `uint8 * uint16 ==> int32`
- etc.

**Comment by blowekamp** [comment]
> Often it is convenient to multiply one image by another, e.g. to mask a segmentation or MRI.

Yes, this is a frequent use case. Fortunately, the [MaskImageFilter](https://simpleitk.org/doxygen/latest/html/classitk_1_1simple_1_1MaskImageFilter.html) is a supports different types and can perform this operations.

> Currently, SimpleITK rasies an exception if the pixel types are different. It would be nice if SimpleITK would automatically cast following similar rules as [standard conversions](https://learn.microsoft.com/en-us/cpp/cpp/standard-conversions?view=msvc-170) in C++, e.g.
> 
> * `float * double ==> double`
> * `float * int16 ==> float`
> * `uint8 * uint16 ==> int32`
> * etc.

This has been considered, and the conclusion was it was best to let the user be in control of how and when Images are promoted in pixel types. Two large factors for this was first that working with large microscopy images having the memory usage unexpectedly double or more due to unavoidable type promotion was undesirable.  The other factor is that instantiating ITK filters for all combinations would be space wise in efficient. It really comes down to what is less likely to a a surprise to the users. Most new users quickly realize that casting is needed, and add it after seeing the error message. Casting is very explicit and predictable.

Related to "masking a segmentation" there is an effort to implement a "mask and assign" filter to enable the following in Python:
```
img[binary_image] = image_or_scalar_value
```

Would that be useful for your situation?

**Comment by dyollb** [comment]
totally understand. Yes, that would help. Thanks!
